ParallelPrimitives/RadixSortKernels.h

```diff
@@ -762,3 +614,131 @@ extern "C" __global__ void ParallelExclusiveScanAllWG( int* gCount, int* gHistog
 
 	gHistogram[blockIdx.x * blockDim.x + threadIdx.x] = blockBuffer[threadIdx.x] + currentGlobalOffset;
 }
+
+template<bool KEY_VALUE_PAIR>
+__device__ void SortImpl( int* gSrcKey, int* gSrcVal, int* gDstKey, int* gDstVal, int* gHistogram, int numberOfInputs, int gNItemsPerWG, const int START_BIT, const int N_WGS_EXECUTED )
+{
+	__shared__ u32 globalOffset[BIN_SIZE];
+	__shared__ u32 localPrefixSum[BIN_SIZE];
+	__shared__ u32 counters[BIN_SIZE];
+
+	__shared__ u32 matchMasks[SORT_NUM_WARPS_PER_BLOCK][BIN_SIZE];
+
+	for( int i = threadIdx.x; i < BIN_SIZE; i += SORT_WG_SIZE )
+	{
+		// Note: The size of gHistogram is always BIN_SIZE * N_WGS_EXECUTED
+		globalOffset[i] = gHistogram[i * N_WGS_EXECUTED + blockIdx.x];
+
+		counters[i] = 0;
+		localPrefixSum[i] = 0;
+	}
+
+	for( int w = 0; w < SORT_NUM_WARPS_PER_BLOCK; ++w )
+	{
+		for( int i = threadIdx.x; i < BIN_SIZE; i += SORT_WG_SIZE )
+		{
+			matchMasks[w][i] = 0;
+		}
+	}
+
+	__syncthreads();
+
+	for( int i = threadIdx.x; i < gNItemsPerWG; i += SORT_WG_SIZE )
+	{
+		const u32 itemIndex = blockIdx.x * gNItemsPerWG + i;
+		if( itemIndex < numberOfInputs )
+		{
+			const auto item = gSrcKey[itemIndex];
+			const u32 bucketIndex = getMaskedBits( item, START_BIT );
+			atomicInc( &localPrefixSum[bucketIndex], 0xFFFFFFFF );
+		}
+	}
+
+	__syncthreads();
+
+	// Compute Prefix Sum
+
+	ldsScanExclusive( localPrefixSum, BIN_SIZE );
+
+	__syncthreads();
+
+	// Reorder
+
+	for( int i = threadIdx.x; i < gNItemsPerWG; i += SORT_WG_SIZE )
+	{
+		const u32 itemIndex = blockIdx.x * gNItemsPerWG + i;
+
+		const auto item = gSrcKey[itemIndex];
+		const u32 bucketIndex = getMaskedBits( item, START_BIT );
+
+		const int warp = threadIdx.x / 32;
+		const int lane = threadIdx.x % 32;
+
+		__syncthreads();
```

Comment:
Note: `syncwarp()` should suffice here but it's not available on HIP.

Reply:
I think someone else suggested but how about making a macro switching these depending on AMD or NV?